### PR TITLE
Renamed some routing APIs to better express their unicast character.

### PR DESCRIPTION
**Pull request by SzymonPobiega**
Changes:
 * `busConfig.Routing()` -> `busConfig.UnicastRouting()`
 * `AddStatic` -> `Add`
 * Moved advanced routing features to `busConfig.AdvancedRouting`
 * Moved filed-based routing config to be available only in *advanced routing* and MSMQ transport config
 * Moved `Publishers` to *advanced routing* and exposed the most common methods directly on `UnicastRouting`.

The new non-advanced API (`busConfig.UnicastRouting()`) is following:

```
RouteToEndpoint(string endpoint);
AddPublisher(string publisherEndpoint, Type eventType);
AddPublisher(string publisherEndpoint, Assembly eventAssembly, string namespace);
```

[Doco PR](https://github.com/Particular/docs.particular.net/pull/1313)

### CODE REVIEW

**Comment by andreasohlund** [comment]
I like it :+1: 


Will you add the


```
if(transport.IsMultiCast && user.HasConfiguredUnicastStuff)
{
    throw new WhatAreYouTryingToDoException();

}

```

in a different pull?


I can't come up with anything(perhaps @johnsimons can) but is there a way for us to only provide this API then the user has selected a unicast transport?

**Comment by SzymonPobiega** [comment]
I've played with various APIs that would enforce unicast/multicast in compile time but failed because:
 * You can pass a `Type` into `UseTransport`
 * The transport might decide if it should use unicast or multicast based on settings

So I decided to rename the `Routing()` into `UnicastRouting()`

**Comment by andreasohlund** [comment]
>So I decided to rename the Routing() into UnicastRouting()

Fair enough, much better than the current naming

File: src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt [hunk 1]
```diff
@@ -2200,8 +2200,8 @@ namespace NServiceBus.Routing
     public class EndpointInstances
     {
         public EndpointInstances() { }
-        public void AddDynamic(System.Func<NServiceBus.Routing.EndpointName, System.Threading.Tasks.Task<System.Collections.Generic.IEnumerable<NServiceBus.Routing.EndpointInstance>>> dynamicRule) { }
-        public void AddStatic(NServiceBus.Routing.EndpointName endpoint, params NServiceBus.Routing.EndpointInstance[] instances) { }
+        public void Add(NServiceBus.Routing.EndpointName endpoint, params NServiceBus.Routing.EndpointInstance[] instances) { }
+        public void AddExternalProvider(System.Func<NServiceBus.Routing.EndpointName, System.Threading.Tasks.Task<System.Collections.Generic.IEnumerable<NServiceBus.Routing.EndpointInstance>>> dynamicRule) { }
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt]: Isn't `Provider` a totally generic word without meaning? How about just call it `Add`, or `AddExternal` or `Add` and `AddDynamic`. Actually `AddDynamic` wasn't too bad

**Comment by andreasohlund** [comment]
Needs a rebase and the doco pull

File: src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt (hunk 1 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt]: ok, I'll leave `AddDynamic`.

**Comment by andreasohlund** [comment]
@danielmarbach added you as a maintainer since you're already involved

**Comment by SzymonPobiega** [comment]
I'll get to that PR, eventually...

**Comment by danielmarbach** [comment]
@SzymonPobiega I could do the rebase for you if you are busy

**Comment by SzymonPobiega** [comment]
@danielmarbach reverted back to `AddDynamic` and added `RouteToEndpoint` directly on `UnicastRoutingSettings`.

**Comment by andreasohlund** [comment]
Needs a rebase due to @danielmarbach cleaning frenzy :)

File: src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/Publishers.cs [hunk 2]
```diff
@@ -12,37 +12,40 @@ public class Publishers
     {
         List<Rule> rules = new List<Rule>();
 
+        internal IEnumerable<PublisherAddress> GetPublisherFor(Type eventType)
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/Publishers.cs]: Should be below the public methods, do a code cleanup?

Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/Publishers.cs]: We need a cleanup on save rule for all contributors

**Comment by SzymonPobiega** [comment]
Hmmm should I do something here?

**Comment by andreasohlund** [comment]
Rebase and a ctrl+E+C on Core/Routing/MessageDrivenSubscriptions/Publishers.cs ?

**Comment by SzymonPobiega** [comment]
:+1:

**Comment by SzymonPobiega** [comment]
Please re-review. I've made some changes after a call where we reviewed the API once again.

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs [hunk 3]
```diff
@@ -105,9 +104,7 @@ public SubscriberB()
             {
                 EndpointSetup<DefaultServer>(c =>
                 {
-                    var publisher = new EndpointName(Conventions.EndpointNamingConvention(typeof(Publisher)));
-                    c.Publishers().AddStatic(publisher, typeof(MyEvent));
-                    c.Routing().EndpointInstances.AddStatic(publisher, new EndpointInstance(publisher));
+                    c.UnicastRouting().AddPublisher(PublisherEdndpoint, typeof(MyEvent));
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs]: Spelling

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs [hunk 4]
```diff
@@ -6,11 +6,12 @@
     using NServiceBus.AcceptanceTesting.Customization;
     using NServiceBus.AcceptanceTests.EndpointTemplates;
     using NServiceBus.AcceptanceTests.ScenarioDescriptors;
-    using NServiceBus.Routing;
     using NUnit.Framework;
 
     public class When_publishing_to_scaled_out_subscribers_on_unicast_transports : NServiceBusAcceptanceTest
     {
+        static string PublisherEdndpoint => Conventions.EndpointNamingConvention(typeof(Publisher));
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs]: Spelling

File: src/NServiceBus.Core/Routing/UnicastRoutingSettings.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        /// </summary>
+        /// <param name="publisherEndpoint">Publisher endpoint.</param>
+        /// <param name="eventAssembly">Assembly containing events.</param>
+        /// <param name="eventNamespace">Optional namespace containing events.</param>
+        public void AddPublisher(string publisherEndpoint, Assembly eventAssembly, string eventNamespace = null)
+        {
+            GetOrCreate<Publishers>().Add(publisherEndpoint, eventAssembly, eventNamespace);
+        }
+
+        T GetOrCreate<T>()
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/UnicastRoutingSettings.cs]: Can't the routing feature just add a default instance of the publishers in the feature constructor?

**Comment by danielmarbach** [comment]
I think a doco PR would really help. I failed to understand why one extension point is called `UnicastRouting` and the other one is called `AdvancedRouting`. The intent for me is not clear.

**Comment by danielmarbach** [comment]
@andreasohlund your thoughts? I'm hesitant to pull this in since I fail to understand the API.

**Comment by andreasohlund** [comment]
>I failed to understand why one extension point is called UnicastRouting and the other one is called AdvancedRouting. 

Unicast makes sense to me since this is not applicable to MultiCast routing. I'm also confused about the advanced part, is that for both multicast and unicast?

File: src/NServiceBus.Core/Routing/UnicastRoutingSettings.cs (hunk 5 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Routing/UnicastRoutingSettings.cs]: It does but this API can be used from `EndpointConfiguration` so before features have chance to register their defaults.

**Comment by SzymonPobiega** [comment]
I've added the WIP doco pull. The `Advanced` part currently has only stuff useful for unicast routing. Would it be better to use `config.UnicastRouting().Advanced.SomeMethodOrProperty`?

**Comment by andreasohlund** [comment]
>Would it be better to use config.UnicastRouting().Advanced.SomeMethodOrProperty?

Sounds better!

**Comment by SzymonPobiega** [comment]
@andreasohlund moved `Advanced` under `Unicast`, rebased and squashed so that there is a commit with all the changes and a separate one only with API approval.

File: src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt [hunk 6]
```diff
@@ -18,6 +18,13 @@ namespace NServiceBus
         Local = 0,
         Remote = 1,
     }
+    public class AdvancedRoutingSettings : NServiceBus.Configuration.AdvanceExtensibility.ExposeSettings
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt]: Should this be `AdvancedUnicastRoutingSettings`?

**Comment by SzymonPobiega** [comment]
Updated `Advanced` -> `Mapping`

**Comment by andreasohlund** [comment]
:+1: to merge, @danielmarbach ?

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs (hunk 4 above)
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs]: @andreasohlund not ok to pull. Still has spelling things to fix

**Comment by danielmarbach** [comment]
Needs a rebase.

**Comment by SzymonPobiega** [comment]
@danielmarbach rebased

**Comment by andreasohlund** [comment]
@SzymonPobiega can you fix the 2 spelling issues that @danielmarbach mentions so we can pull?

**Comment by SzymonPobiega** [comment]
@andreasohlund done.

**Pull request merged**